File: codes/java/648.单词替换.java
class Trie {
	Trie[] children;
	boolean isEnd;

	public Trie() {
		this.children = new Trie[26];
		this.isEnd = false;
	}

	public void insert(String word) {
		Trie node = this;
		for (char ch : word.toCharArray()) {
			if (node.children[ch - 'a'] == null) {
				node.children[ch - 'a'] = new Trie();
			}
			node = node.children[ch - 'a'];
		}
		node.isEnd = true;
	}
}

class Solution {
	public String replaceWords(List<String> dictionary, String sentence) {
		// 首先对词典构建字典树
		Trie trie = new Trie();
		for (String root : dictionary) {
			trie.insert(root);
		}

		StringBuilder ans = new StringBuilder();
		// 接着分割sentance，通过空格
		String[] words = sentence.split(" ");
		for (String word : words) {
			// 对每个单词查找前缀
			Trie node = trie;
			boolean find_prefix = false;
			String prefix = "";
			for (char ch : word.toCharArray()) {
				if (node.children[ch - 'a'] != null) {
					prefix += ch;
					if (node.children[ch - 'a'].isEnd) {
						find_prefix = true;
						break;
					}
					node = node.children[ch - 'a'];
				} else {
					break;
				}
			}
			if (find_prefix) {
				ans.append(prefix);
			} else {
				ans.append(word);
			}
			ans.append(" ");
		}
		return ans.toString().trim();
	}
}

File: codes/python/648.单词替换.py
class Trie:
    def __init__(self):
        self.children = [None] * 26
        self.isEnd = False

    def insert(self, word: str):
        node = self
        for ch in word:
            if not node.children[ord(ch) - ord('a')]:
                node.children[ord(ch) - ord('a')] = Trie()
            node = node.children[ord(ch) - ord('a')]
        node.isEnd = True

class Solution:
    def replaceWords(self, dictionary: List[str], sentence: str) -> str:
        trie = Trie()
        for root in dictionary:
            trie.insert(root)
        
        ans = []
        words = sentence.split()
        for word in words:
            node = trie
            find_prefix = False
            prefix = ""
            for ch in word:
                if node.children[ord(ch) - ord('a')]:
                    prefix += ch
                    if node.children[ord(ch) - ord('a')].isEnd:
                        find_prefix = True
                        break
                    node = node.children[ord(ch) - ord('a')]
                else:
                    break
            if find_prefix:
                ans.append(prefix)
            else:
                ans.append(word)
        return " ".join(ans)
        